File: 4.8-bc/go/stage-3-hexagonal/shared/events/bus.go
// Package events — мінімальний in-memory event bus для cross-BC комунікації.
// BC-публікатор не знає про підписників. BC-підписник реєструє handler у своєму infra/events/.
package events

import (
	"context"
	"sync"
)

type Event interface {
	Name() string
}

type Handler func(ctx context.Context, e Event)

type Bus struct {
	mu       sync.RWMutex
	handlers map[string][]Handler
}

func NewBus() *Bus {
	return &Bus{handlers: make(map[string][]Handler)}
}

func (b *Bus) Subscribe(name string, h Handler) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.handlers[name] = append(b.handlers[name], h)
}

func (b *Bus) Publish(ctx context.Context, e Event) {
	b.mu.RLock()
	hs := append([]Handler(nil), b.handlers[e.Name()]...)
	b.mu.RUnlock()
	for _, h := range hs {
		h(ctx, e)
	}
}

File: 4.8-bc/ts/stage-3-hexagonal/src/shared/events/bus.ts
// shared/events/bus — мінімальний in-memory event bus для cross-BC комунікації.
// BC-публікатор не знає про підписників. BC-підписник реєструє handler у своєму infra/events/.

export interface DomainEvent {
  readonly name: string;
}

export type Handler<E extends DomainEvent> = (e: E) => void | Promise<void>;

export class EventBus {
  private readonly handlers: Map<string, Handler<DomainEvent>[]> = new Map();

  subscribe<E extends DomainEvent>(name: string, h: Handler<E>): void {
    const list = this.handlers.get(name) ?? [];
    // Cast safely: handler bound to a specific event name receives that event.
    list.push(h as unknown as Handler<DomainEvent>);
    this.handlers.set(name, list);
  }

  async publish<E extends DomainEvent>(e: E): Promise<void> {
    const list = this.handlers.get(e.name);
    if (!list) return;
    for (const h of list) {
      await h(e);
    }
  }
}

export function newBus(): EventBus {
  return new EventBus();
}
